python/tskit/trees.py

```diff
@@ -9291,25 +9290,156 @@ def ibd_segments(
+        """
+
+        if sample_sets is None:
+            sample_sets = [list(self.samples())]
+
+        if indexes is None:
+            sample_set_ids = np.arange(len(sample_sets))
+            indexes = list(itertools.combinations_with_replacement(sample_set_ids, 2))
+
+        if windows is None:
+            windows = np.array([0.0, self.sequence_length])
+
+        if not (windows[0] == 0.0 and windows[-1] == self.sequence_length):
+            raise ValueError("First and last window breaks must be sequence boundary")
+        if not np.all(np.diff(windows) > 0):
+            raise ValueError("Window breaks must be strictly increasing")
+        for s in sample_sets:
+            if len(s) == 0:
+                raise ValueError("Sample set is length zero")
+            if not np.all(np.bitwise_and(self.nodes_flags[s], tskit.NODE_IS_SAMPLE)):
+                raise ValueError("Sample set contains non-sample node")
+        for i in indexes:
+            if not len(i) == 2:
+                raise ValueError("Sample set indexes must be length two")
+            if not (min(i) >= 0 and max(i) < len(sample_sets)):
+                raise ValueError("Sample set index is out of bounds")
+
+        num_nodes = self.num_nodes
+        num_edges = self.num_edges
+        num_windows = windows.size - 1
+        num_sample_sets = len(sample_sets)
+        num_indexes = len(indexes)
+
+        edges_child = self.edges_child
+        edges_parent = self.edges_parent
+        insert_index = self.indexes_edge_insertion_order
+        remove_index = self.indexes_edge_removal_order
+        insert_position = self.edges_left[insert_index]
+        remove_position = self.edges_right[remove_index]
+        sequence_length = self.sequence_length
+
+        nodes_parent = np.full(num_nodes, tskit.NULL)
+        sample_counts = np.zeros((num_nodes, num_sample_sets))
+        coalescing_pairs = np.zeros((num_windows, num_nodes, num_indexes))
+
+        for i, s in enumerate(sample_sets):
+            sample_counts[s, i] = 1
+
+        position = 0.0
+        w, a, b = 0, 0, 0
+        while position < sequence_length:
+            remainder = sequence_length - position
+
+            while b < num_edges and remove_position[b] == position:  # edges out
+                e = remove_index[b]
+                p = edges_parent[e]
+                c = edges_child[e]
+                nodes_parent[c] = tskit.NULL
+                inside = sample_counts[c]
+                while p != tskit.NULL:
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] -= weight * remainder
+                    c, p = p, nodes_parent[p]
+                p = edges_parent[e]
+                while p != tskit.NULL:
+                    sample_counts[p] -= inside
```

Comment:
I think this can move up three lines, and eliminate this second while loop?

Reply:
This won't work, unfortunately -- the first while loop needs to use `sample_counts[p]` before it's modified, for all `p` on the path from the first child to the root. Only after these are updated can `sample_counts` be modified.